File: swift/Sources/MaterialColorUtilities/Contrast/Contrast.swift
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/// Utility methods for calculating contrast given two colors, or calculating a
/// color given one color and a contrast ratio.
///
/// Contrast ratio is calculated using XYZ's Y. When linearized to match human
/// perception, Y becomes HCT's tone and L*a*b*'s' L*. Informally, this is the
/// lightness of a color.
///
/// Methods refer to tone, T in the the HCT color space.
/// Tone is equivalent to L* in the L*a*b* color space, or L in the LCH color
/// space.
public class Contrast {
  /// Returns a contrast ratio, which ranges from 1 to 21.
  /// [toneA] Tone between 0 and 100. Values outside will be clamped.
  /// [toneB] Tone between 0 and 100. Values outside will be clamped.
  public static func ratioOfTones(_ toneA: Double, _ toneB: Double) -> Double {
    let toneA = MathUtils.clampDouble(0, 100, toneA)
    let toneB = MathUtils.clampDouble(0, 100, toneB)
    return ratioOfYs(ColorUtils.yFromLstar(toneA), ColorUtils.yFromLstar(toneB))
  }

  static private func ratioOfYs(_ y1: Double, _ y2: Double) -> Double {
    let lighter = y1 > y2 ? y1 : y2
    let darker = lighter == y2 ? y1 : y2
    return (lighter + 5.0) / (darker + 5.0)
  }

  /// Returns a tone >= [tone] that ensures [ratio].
  /// Return value is between 0 and 100.
  /// Returns -1 if [ratio] cannot be achieved with [tone].
  ///
  /// [tone] Tone return value must contrast with.
  /// Range is 0 to 100. Invalid values will result in -1 being returned.
  /// [ratio] Contrast ratio of return value and [tone].
  /// Range is 1 to 21, invalid values have undefined behavior.
  public static func lighter(tone: Double, ratio: Double) -> Double {
    if tone < 0 || tone > 100 {
      return -1
    }

    let darkY = ColorUtils.yFromLstar(tone)
    let lightY = ratio * (darkY + 5.0) - 5.0
    let realContrast = ratioOfYs(lightY, darkY)
    let delta = abs(realContrast - ratio)
    if realContrast < ratio && delta > 0.04 {
      return -1
    }

    // Ensure gamut mapping, which requires a 'range' on tone, will still result
    // the correct ratio by darkening slightly.
    let returnValue = ColorUtils.lstarFromY(lightY) + 0.4
    if returnValue < 0 || returnValue > 100 {
      return -1
    }
    return returnValue
  }

  /// Returns a tone <= [tone] that ensures [ratio].
  /// Return value is between 0 and 100.
  /// Returns -1 if [ratio] cannot be achieved with [tone].
  ///
  /// [tone] Tone return value must contrast with.
  /// Range is 0 to 100. Invalid values will result in -1 being returned.
  /// [ratio] Contrast ratio of return value and [tone].
  /// Range is 1 to 21, invalid values have undefined behavior.
  public static func darker(tone: Double, ratio: Double) -> Double {
    if tone < 0 || tone > 100 {
      return -1
    }

    let lightY = ColorUtils.yFromLstar(tone)
    let darkY = ((lightY + 5.0) / ratio) - 5.0
    let realContrast = ratioOfYs(lightY, darkY)

    let delta = abs(realContrast - ratio)
    if realContrast < ratio && delta > 0.04 {
      return -1
    }

    // Ensure gamut mapping, which requires a 'range' on tone, will still result
    // the correct ratio by darkening slightly.
    let returnValue = ColorUtils.lstarFromY(darkY) - 0.4
    if returnValue < 0 || returnValue > 100 {
      return -1
    }
    return returnValue
  }

  /// Returns a tone >= [tone] that ensures [ratio].
  /// Return value is between 0 and 100.
  /// Returns 100 if [ratio] cannot be achieved with [tone].
  ///
  /// This method is unsafe because the returned value is guaranteed to be in
  /// bounds for tone, i.e. between 0 and 100. However, that value may not reach
  /// the [ratio] with [tone]. For example, there is no color lighter than T100.
  ///
  /// [tone] Tone return value must contrast with.
  /// Range is 0 to 100. Invalid values will result in 100 being returned.
  /// [ratio] Desired contrast ratio of return value and tone parameter.
  /// Range is 1 to 21, invalid values have undefined behavior.
  public static func lighterUnsafe(tone: Double, ratio: Double) -> Double {
    let lighterSafe = lighter(tone: tone, ratio: ratio)
    return lighterSafe < 0 ? 100 : lighterSafe
  }

  /// Returns a tone <= [tone] that ensures [ratio].
  /// Return value is between 0 and 100.
  /// Returns 0 if [ratio] cannot be achieved with [tone].
  ///
  /// This method is unsafe because the returned value is guaranteed to be in
  /// bounds for tone, i.e. between 0 and 100. However, that value may not reach
  /// the [ratio] with [tone]. For example, there is no color darker than T0.
  ///
  /// [tone] Tone return value must contrast with.
  /// Range is 0 to 100. Invalid values will result in 0 being returned.
  /// [ratio] Desired contrast ratio of return value and tone parameter.
  /// Range is 1 to 21, invalid values have undefined behavior.
  public static func darkerUnsafe(tone: Double, ratio: Double) -> Double {
    let darkerSafe = darker(tone: tone, ratio: ratio)
    return darkerSafe < 0 ? 0 : darkerSafe
  }
}

File: go/contrast/contrast.go
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package contrast provides utilities for calculating and ensuring color contrast ratios.
package contrast

import (
	"math"

	"github.com/material-foundation/material-color-utilities/go/utils"
)

// Standard contrast ratio constants
const (
	RatioMin = 1.0  // Minimum contrast ratio of two colors
	RatioMax = 21.0 // Maximum contrast ratio of two colors
	Ratio30  = 3.0  // WCAG AA normal text
	Ratio45  = 4.5  // WCAG AA large text
	Ratio70  = 7.0  // WCAG AAA
)

// Internal constants for contrast calculations
const (
	contrastRatioEpsilon       = 0.04 // Tolerance for contrast ratio calculations
	luminanceGamutMapTolerance = 0.4  // Tolerance for gamut mapping
)

// RatioOfYs calculates the contrast ratio of two Y values (relative luminance).
// This is the standard WCAG contrast ratio calculation.
func RatioOfYs(y1, y2 float64) float64 {
	lighter := math.Max(y1, y2)
	var darker float64
	if lighter == y2 {
		darker = y1
	} else {
		darker = y2
	}
	return (lighter + 5.0) / (darker + 5.0)
}

// RatioOfTones calculates the contrast ratio of two tones (T in HCT, L* in L*a*b*).
func RatioOfTones(t1, t2 float64) float64 {
	return RatioOfYs(utils.YFromLstar(t1), utils.YFromLstar(t2))
}

// Lighter returns a tone >= the input tone that ensures the given contrast ratio.
// Returns -1 if the ratio cannot be achieved.
func Lighter(tone, ratio float64) float64 {
	if tone < 0.0 || tone > 100.0 {
		return -1.0
	}

	// Invert the contrast ratio equation to determine lighter Y given a ratio and darker Y
	darkY := utils.YFromLstar(tone)
	lightY := ratio*(darkY+5.0) - 5.0
	if lightY < 0.0 || lightY > 100.0 {
		return -1.0
	}

	realContrast := RatioOfYs(lightY, darkY)
	delta := math.Abs(realContrast - ratio)
	if realContrast < ratio && delta > contrastRatioEpsilon {
		return -1.0
	}

	returnValue := utils.LstarFromY(lightY) + luminanceGamutMapTolerance
	if returnValue < 0 || returnValue > 100 {
		return -1.0
	}
	return returnValue
}

// LighterUnsafe returns a tone >= the input tone that ensures the given contrast ratio.
// Unlike Lighter, this function does not validate the result and returns 100 if
// the ratio cannot be achieved.
func LighterUnsafe(tone, ratio float64) float64 {
	lighter := Lighter(tone, ratio)
	if lighter < 0 {
		return 100.0
	}
	return lighter
}

// Darker returns a tone <= the input tone that ensures the given contrast ratio.
// Returns -1 if the ratio cannot be achieved.
func Darker(tone, ratio float64) float64 {
	if tone < 0.0 || tone > 100.0 {
		return -1.0
	}

	// Invert the contrast ratio equation to determine darker Y given a ratio and lighter Y
	lightY := utils.YFromLstar(tone)
	darkY := ((lightY + 5.0) / ratio) - 5.0
	if darkY < 0.0 || darkY > 100.0 {
		return -1.0
	}

	realContrast := RatioOfYs(lightY, darkY)
	delta := math.Abs(realContrast - ratio)
	if realContrast < ratio && delta > contrastRatioEpsilon {
		return -1.0
	}

	returnValue := utils.LstarFromY(darkY) - luminanceGamutMapTolerance
	if returnValue < 0 || returnValue > 100 {
		return -1.0
	}
	return returnValue
}

// DarkerUnsafe returns a tone <= the input tone that ensures the given contrast ratio.
// Unlike Darker, this function does not validate the result and returns 0 if
// the ratio cannot be achieved.
func DarkerUnsafe(tone, ratio float64) float64 {
	darker := Darker(tone, ratio)
	if darker < 0 {
		return 0.0
	}
	return darker
}
